File: core/src/main/kotlin/com/pokerface/core/Card.kt
package com.pokerface.core

enum class Suit(val symbol: String, val label: String) {
    SPADES("♠", "스페이드"),
    HEARTS("♥", "하트"),
    DIAMONDS("♦", "다이아"),
    CLUBS("♣", "클럽");

    val isRed: Boolean get() = this == HEARTS || this == DIAMONDS

    companion object {
        fun fromChar(c: Char): Suit = when (c.lowercaseChar()) {
            's' -> SPADES
            'h' -> HEARTS
            'd' -> DIAMONDS
            'c' -> CLUBS
            else -> throw IllegalArgumentException("알 수 없는 무늬 문자: $c")
        }
    }
}

enum class Rank(val value: Int, val label: String) {
    TWO(2, "2"), THREE(3, "3"), FOUR(4, "4"), FIVE(5, "5"), SIX(6, "6"),
    SEVEN(7, "7"), EIGHT(8, "8"), NINE(9, "9"), TEN(10, "10"),
    JACK(11, "J"), QUEEN(12, "Q"), KING(13, "K"), ACE(14, "A");

    companion object {
        fun fromChar(c: Char): Rank = when (c.uppercaseChar()) {
            '2' -> TWO; '3' -> THREE; '4' -> FOUR; '5' -> FIVE; '6' -> SIX
            '7' -> SEVEN; '8' -> EIGHT; '9' -> NINE; 'T' -> TEN
            'J' -> JACK; 'Q' -> QUEEN; 'K' -> KING; 'A' -> ACE
            else -> throw IllegalArgumentException("알 수 없는 랭크 문자: $c")
        }

        fun fromLabel(label: String): Rank = when (label.uppercase()) {
            "10", "T" -> TEN
            else -> fromChar(label[0])
        }
    }
}

data class Card(val rank: Rank, val suit: Suit) : Comparable<Card> {

    override fun compareTo(other: Card): Int = rank.value - other.rank.value

    override fun toString(): String = "${rank.label}${suit.symbol}"

    /** 예: "As", "Td", "9h", "10c" */
    fun toCode(): String = "${rank.label}${suit.name.first().lowercaseChar()}"

    companion object {
        /** "As" / "Td" / "9h" / "10c" 같은 표기를 파싱한다. */
        fun fromCode(code: String): Card {
            val trimmed = code.trim()
            require(trimmed.length in 2..3) { "잘못된 카드 코드: $code" }
            val rankPart = trimmed.dropLast(1)
            val suitChar = trimmed.last()
            return Card(Rank.fromLabel(rankPart), Suit.fromChar(suitChar))
        }
    }
}

File: web/src/poker/card.ts
export enum Suit {
  SPADES = "SPADES",
  HEARTS = "HEARTS",
  DIAMONDS = "DIAMONDS",
  CLUBS = "CLUBS",
}

export const SUIT_SYMBOL: Record<Suit, string> = {
  [Suit.SPADES]: "♠",
  [Suit.HEARTS]: "♥",
  [Suit.DIAMONDS]: "♦",
  [Suit.CLUBS]: "♣",
};

export const SUIT_LABEL_KO: Record<Suit, string> = {
  [Suit.SPADES]: "스페이드",
  [Suit.HEARTS]: "하트",
  [Suit.DIAMONDS]: "다이아",
  [Suit.CLUBS]: "클럽",
};

export const ALL_SUITS: Suit[] = [Suit.SPADES, Suit.HEARTS, Suit.DIAMONDS, Suit.CLUBS];

export function isRedSuit(suit: Suit): boolean {
  return suit === Suit.HEARTS || suit === Suit.DIAMONDS;
}

export enum Rank {
  TWO = 2,
  THREE = 3,
  FOUR = 4,
  FIVE = 5,
  SIX = 6,
  SEVEN = 7,
  EIGHT = 8,
  NINE = 9,
  TEN = 10,
  JACK = 11,
  QUEEN = 12,
  KING = 13,
  ACE = 14,
}

export const ALL_RANKS: Rank[] = [
  Rank.TWO, Rank.THREE, Rank.FOUR, Rank.FIVE, Rank.SIX, Rank.SEVEN, Rank.EIGHT,
  Rank.NINE, Rank.TEN, Rank.JACK, Rank.QUEEN, Rank.KING, Rank.ACE,
];

export function rankLabel(rank: Rank): string {
  switch (rank) {
    case Rank.ACE: return "A";
    case Rank.KING: return "K";
    case Rank.QUEEN: return "Q";
    case Rank.JACK: return "J";
    default: return String(rank);
  }
}

export function rankFromLabel(label: string): Rank | null {
  const cleaned = label.trim().toUpperCase();
  switch (cleaned) {
    case "A": return Rank.ACE;
    case "K": return Rank.KING;
    case "Q": return Rank.QUEEN;
    case "J": return Rank.JACK;
    case "10": case "T": return Rank.TEN;
    case "9": return Rank.NINE;
    case "8": return Rank.EIGHT;
    case "7": return Rank.SEVEN;
    case "6": return Rank.SIX;
    case "5": return Rank.FIVE;
    case "4": return Rank.FOUR;
    case "3": return Rank.THREE;
    case "2": return Rank.TWO;
    default: return null;
  }
}

export interface Card {
  rank: Rank;
  suit: Suit;
}

export function makeCard(rank: Rank, suit: Suit): Card {
  return { rank, suit };
}

export function cardCode(card: Card): string {
  return `${rankLabel(card.rank)}${card.suit[0]}`;
}

export function cardsEqual(a: Card, b: Card): boolean {
  return a.rank === b.rank && a.suit === b.suit;
}

export function cardKey(card: Card): string {
  return `${card.rank}-${card.suit}`;
}

/** "As" / "Td" / "9h" / "10c" 같은 표기를 파싱한다. 테스트 편의용. */
export function cardFromCode(code: string): Card {
  const trimmed = code.trim();
  const rankPart = trimmed.slice(0, -1);
  const suitChar = trimmed.slice(-1).toLowerCase();
  const rank = rankFromLabel(rankPart);
  if (rank === null) throw new Error(`잘못된 카드 코드: ${code}`);
  const suit = ALL_SUITS.find((s) => s[0].toLowerCase() === suitChar);
  if (!suit) throw new Error(`잘못된 카드 코드: ${code}`);
  return { rank, suit };
}
